File: 2021/09/03/NumberComplement.java
/**
 * https://leetcode.com/problems/number-complement/
 * @author Andrew Jarombek
 * @since 9/3/2021
 */

public class NumberComplement {
    public static int findComplement(int num) {
        String binary = Integer.toString(num, 2);

        int result = 0;
        for (int i = binary.length() - 1; i >= 0; i--) {
            if (binary.charAt(i) == '0') {
                result += Math.pow(2, Math.abs(binary.length() - i - 1));
            }
        }

        return result;
    }

    public static void main(String[] args) {
        assert findComplement(5) == 2;
    }
}

File: 2023/06/30/go/number_complement.go
/**
 * https://leetcode.com/problems/island-perimeter/
 * Author: Andrew Jarombek
 * Date: 6/30/2023
 */

package _go

func findComplement(num int) int {
	result := 0
	power := 1

	for num > 0 {
		if num%2 == 0 {
			result += power
		}

		power <<= 1
		num >>= 1
	}

	return result
}
